### PR TITLE
Implement config query tool

### PR DESCRIPTION
**Pull request by kristian-lesko**
Add the QueryTool component that'll enable various operations
over the entity config, such as checking membership of entity
in another entity, or checking if a required security label's
present in an entity as required by another entities' config.

For now, nested membership check is implemented only; labels'
checks are going to be implemented soon in a separate commit.

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/packages/job/freeipa-manager-docker-tox-zuul-el7/79/ : SUCCESS in 49s
- https://checklist.intgdc.com/job/compliance/job/freeipa-manager-opensource-repository-check/17/ : SUCCESS in 9s

File: ipamanager/tools/query_tool.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        """
+        result = self.graph.get(entity, set())
+        if result:
+            self.lg.debug('Membership for %s already calculated', entity)
+            return result
+        self.lg.debug('Calculating membership graph for %s', entity)
+        memberof = entity.data_repo.get('memberOf', {})
+        for target_type, targets in memberof.iteritems():
+            for target in targets:
+                target_entity = find_entity(self.entities, target_type, target)
```
Comment (pbenas) [inline ipamanager/tools/query_tool.py]: Slightly confused by naming here. `target_type` is obvious, but what's the difference between `target` and `target_entity` here?

File: ipamanager/tools/query_tool.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+                else:
+                    self.predecessors[target_entity] = [entity]
+                result.update(self.build_graph(target_entity))
+        self.lg.debug('Found %d target entities for %s', len(result), entity)
+        self.graph[entity] = result
+        return result
+
+    def check_membership(self, entity, target):
+        """
+        Check if `entity` is a member of `target`.
```
Comment (pbenas) [inline ipamanager/tools/query_tool.py]: I don't like the word `target` much.`member` is quite clear what the relationship is. I would rename "entity" to "memter" or "candidate" and "target" to "predecessor" or even to "entity".

File: ipamanager/tools/query_tool.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        member_entities = self._resolve_entities(members)
+        target_entities = self._resolve_entities(targets)
+        for entity in member_entities:
+            for target in target_entities:
+                self.check_membership(entity, target)
+
+    def _construct_path(self, target, member):
+        """
+        Find all paths leading from `member` to `target` entity.
+        The paths are constructed backwards, based on the `predecessor`
```
Comment (pbenas) [inline ipamanager/tools/query_tool.py]: other naming suggestion, does `predecessor` imply a `direct predecessor`? Maybe that's what a `parent` is.

File: ipamanager/tools/query_tool.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    actions = parser.add_subparsers(help='query action to execute')
+
+    member = actions.add_parser('member', parents=[common])
+    member.add_argument(
+        '-m', '--members', nargs='+', type=_entity_type, default=[],
+        required=True, help='member entities (type:name format)')
+    member.add_argument(
+        '-t', '--targets', nargs='+', type=_entity_type, default=[],
+        required=True, help='target entities (type:name format)')
+    member.set_defaults(action='member')
```
Comment (pbenas) [inline ipamanager/tools/query_tool.py]: I'm not sure whether I'd be able to use it right away just after running `--help`. Please consider documenting it on some simple example in README.md

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/packages/job/freeipa-manager-docker-tox-zuul-el7/82/ : SUCCESS in 46s
- https://checklist.intgdc.com/job/compliance/job/freeipa-manager-merge-compliance-check/24/ : SUCCESS in 31s

**Pull request merged**